// @vibeflowai/convex-mcp/dist/lib/prompts.js
import { normalizeToolArgs } from "./schema.js";
const EMPTY_OBJECT_VALIDATOR = {
    type: "object",
    value: {},
};
function toPromptArguments(shape) {
    const entries = Object.entries(shape);
    if (entries.length === 0) {
        return undefined;
    }
    return entries.map(([name, schema]) => ({
        name,
        description: schema.description,
        required: !schema.safeParse(undefined).success,
    }));
}
export function normalizePrompt(name, definition) {
    const path = name.split(".").filter(Boolean);
    if (path.length === 0) {
        throw new Error(`Prompt name "${name}" must be a dot-delimited path.`);
    }
    const { inputShape } = normalizeToolArgs(EMPTY_OBJECT_VALIDATOR, definition.args);
    return {
        name,
        title: definition.title,
        description: definition.description,
        arguments: toPromptArguments(inputShape),
        inputShape,
        handler: definition.handler,
    };
}
function isRawPromptDefinition(value) {
    return !!value && typeof value === "object" && "handler" in value;
}
export function flattenPrompts(prompts = {}, prefix = []) {
    const flattened = [];
    for (const [key, value] of Object.entries(prompts)) {
        const nextPrefix = [...prefix, ...key.split(".").filter(Boolean)];
        if (isRawPromptDefinition(value)) {
            flattened.push([nextPrefix.join("."), value]);
            continue;
        }
        flattened.push(...flattenPrompts(value, nextPrefix));
    }
    return flattened;
}
export function prompt(options, handler) {
    return {
        title: options.title,
        description: options.description,
        args: options.args,
        handler,
    };
}
export function promptMessage(role, content) {
    return { role, content };
}
export function textContent(text) {
    return {
        type: "text",
        text,
    };
}
export function userText(text) {
    return promptMessage("user", textContent(text));
}
export function assistantText(text) {
    return promptMessage("assistant", textContent(text));
}
export function promptResult(messages, description) {
    return {
        description,
        messages,
    };
}
export function toPromptResult(result) {
    if (Array.isArray(result)) {
        return {
            messages: result,
        };
    }
    if (isPromptMessage(result)) {
        return {
            messages: [result],
        };
    }
    return result;
}
function isPromptMessage(value) {
    return !!value && typeof value === "object" && !Array.isArray(value) &&
        "role" in value && "content" in value;
}
