// @vibeflowai/convex-mcp/dist/lib/resources.js
import { UriTemplate } from "@modelcontextprotocol/sdk/shared/uriTemplate.js";
import { normalizeToolArgs } from "./schema.js";
const EMPTY_OBJECT_VALIDATOR = {
    type: "object",
    value: {},
};
function assertReadKind(kind, label, name) {
    if (!kind) {
        throw new Error(`${label} "${name}" uses an api/internal function reference, so you must provide "kind".`);
    }
    return kind;
}
function templateValidator(uriTemplate) {
    return {
        type: "object",
        value: Object.fromEntries(uriTemplate.variableNames.map((name) => [
            name,
            {
                fieldType: { type: "string" },
                optional: false,
            },
        ])),
    };
}
function isRawResourceDefinition(value) {
    return !!value && typeof value === "object" && "ref" in value && "uri" in value;
}
function isRawResourceTemplateDefinition(value) {
    return !!value && typeof value === "object" && "ref" in value &&
        "uriTemplate" in value;
}
export function normalizeResource(name, definition) {
    const path = name.split(".").filter(Boolean);
    if (path.length === 0) {
        throw new Error(`Resource name "${name}" must be a dot-delimited path.`);
    }
    if (!definition.uri) {
        throw new Error(`Resource "${name}" must define a non-empty "uri".`);
    }
    return {
        name,
        kind: assertReadKind(definition.kind, "Resource", name),
        ref: definition.ref,
        uri: definition.uri,
        title: definition.title,
        description: definition.description,
        mimeType: definition.mimeType,
        annotations: definition.annotations,
        size: definition.size,
    };
}
export function normalizeResourceTemplate(name, definition) {
    const path = name.split(".").filter(Boolean);
    if (path.length === 0) {
        throw new Error(`Resource template name "${name}" must be a dot-delimited path.`);
    }
    const matcher = new UriTemplate(definition.uriTemplate);
    const { inputShape } = normalizeToolArgs(templateValidator(matcher), definition.params);
    return {
        name,
        kind: assertReadKind(definition.kind, "Resource template", name),
        ref: definition.ref,
        uriTemplate: definition.uriTemplate,
        matcher,
        title: definition.title,
        description: definition.description,
        mimeType: definition.mimeType,
        annotations: definition.annotations,
        inputShape,
    };
}
export function flattenResources(resources = {}, prefix = []) {
    const flattened = [];
    for (const [key, value] of Object.entries(resources)) {
        const nextPrefix = [...prefix, ...key.split(".").filter(Boolean)];
        if (isRawResourceDefinition(value)) {
            flattened.push([nextPrefix.join("."), value]);
            continue;
        }
        flattened.push(...flattenResources(value, nextPrefix));
    }
    return flattened;
}
export function flattenResourceTemplates(resources = {}, prefix = []) {
    const flattened = [];
    for (const [key, value] of Object.entries(resources)) {
        const nextPrefix = [...prefix, ...key.split(".").filter(Boolean)];
        if (isRawResourceTemplateDefinition(value)) {
            flattened.push([nextPrefix.join("."), value]);
            continue;
        }
        flattened.push(...flattenResourceTemplates(value, nextPrefix));
    }
    return flattened;
}
export async function invokeResource(ctx, resource, args) {
    switch (resource.kind) {
        case "query":
            return await ctx.runQuery(resource.ref, args);
        case "action":
            return await ctx.runAction(resource.ref, args);
    }
}
export function findResourceByUri(resources, resourceTemplates, uri) {
    for (const resource of resources.values()) {
        if (resource.uri === uri) {
            return {
                kind: "resource",
                definition: resource,
                params: {},
            };
        }
    }
    for (const resourceTemplate of resourceTemplates.values()) {
        const params = resourceTemplate.matcher.match(uri);
        if (params) {
            return {
                kind: "template",
                definition: resourceTemplate,
                params,
            };
        }
    }
    return null;
}
export function resource(ref, options) {
    return {
        ref,
        kind: options.kind,
        uri: options.uri,
        title: options.title,
        description: options.description,
        mimeType: options.mimeType,
        annotations: options.annotations,
        size: options.size,
    };
}
export function resourceTemplate(ref, options) {
    return {
        ref,
        kind: options.kind,
        uriTemplate: options.uriTemplate,
        title: options.title,
        description: options.description,
        mimeType: options.mimeType,
        annotations: options.annotations,
        params: options.params,
    };
}
export function textResource(text, options = {}) {
    return {
        type: "text",
        text,
        ...options,
    };
}
export function jsonResource(value, options = {}) {
    return {
        type: "json",
        value,
        ...options,
    };
}
export function blobResource(blob, mimeType, options = {}) {
    return {
        type: "blob",
        blob,
        mimeType,
        ...options,
    };
}
export function resourceResult(...contents) {
    return {
        contents,
    };
}
