// @vibeflowai/convex-mcp/dist/lib/result.js
function formatStructuredValue(value) {
    return JSON.stringify(value, null, 2);
}
export function toCallToolResult(value) {
    if (typeof value === "string") {
        return {
            content: [{ type: "text", text: value }],
        };
    }
    if (value !== null &&
        typeof value === "object" &&
        !Array.isArray(value)) {
        return {
            content: [{ type: "text", text: formatStructuredValue(value) }],
            structuredContent: value,
        };
    }
    return {
        content: [{ type: "text", text: formatStructuredValue(value) }],
        structuredContent: { result: value },
    };
}
export function toErrorToolResult(error) {
    const message = error instanceof Error ? error.message : "Unknown MCP tool error";
    return {
        content: [{ type: "text", text: message }],
        isError: true,
    };
}
function isResourceContentInput(value) {
    return !!value && typeof value === "object" && !Array.isArray(value) &&
        "type" in value &&
        (value.type === "text" || value.type === "blob" || value.type === "json");
}
function isRawResourceResult(value) {
    return !!value && typeof value === "object" && !Array.isArray(value) &&
        "contents" in value &&
        Array.isArray(value.contents);
}
function isReadResourceResult(value) {
    return isRawResourceResult(value) &&
        value.contents.every((item) => !!item && typeof item === "object" && "uri" in item &&
            ("text" in item || "blob" in item));
}
function jsonMimeType(defaultMimeType) {
    return defaultMimeType?.includes("json")
        ? defaultMimeType
        : "application/json";
}
function toReadResourceContents(value, defaults) {
    switch (value.type) {
        case "text":
            return {
                uri: value.uri ?? defaults.uri,
                mimeType: value.mimeType ?? defaults.mimeType ?? "text/plain",
                text: value.text,
                ...(value._meta ? { _meta: value._meta } : {}),
            };
        case "blob":
            return {
                uri: value.uri ?? defaults.uri,
                mimeType: value.mimeType ?? defaults.mimeType,
                blob: value.blob,
                ...(value._meta ? { _meta: value._meta } : {}),
            };
        case "json":
            return {
                uri: value.uri ?? defaults.uri,
                mimeType: value.mimeType ?? jsonMimeType(defaults.mimeType),
                text: formatStructuredValue(value.value),
                ...(value._meta ? { _meta: value._meta } : {}),
            };
    }
}
export function toReadResourceResult(value, defaults) {
    if (isReadResourceResult(value)) {
        return value;
    }
    if (isRawResourceResult(value)) {
        return {
            contents: value.contents.map((content) => isResourceContentInput(content)
                ? toReadResourceContents(content, defaults)
                : {
                    uri: defaults.uri,
                    mimeType: defaults.mimeType ?? "text/plain",
                    text: formatStructuredValue(content),
                }),
        };
    }
    if (Array.isArray(value) && value.every(isResourceContentInput)) {
        const contents = value;
        return {
            contents: contents.map((content) => toReadResourceContents(content, defaults)),
        };
    }
    if (isResourceContentInput(value)) {
        return {
            contents: [toReadResourceContents(value, defaults)],
        };
    }
    if (typeof value === "string") {
        return {
            contents: [
                {
                    uri: defaults.uri,
                    mimeType: defaults.mimeType ?? "text/plain",
                    text: value,
                },
            ],
        };
    }
    return {
        contents: [
            {
                uri: defaults.uri,
                mimeType: jsonMimeType(defaults.mimeType),
                text: formatStructuredValue(value),
            },
        ],
    };
}
